[WIP] Introduce Map

mcpl-sympy opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Derived from #19750

#### Brief description of what is fixed or changed

Map class is introduced. This class is function as object, not class as `Function` is.

#### Other comments



#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- map
    - map.py is introduced
<!-- END RELEASE NOTES -->

JSS95 committed:
Extract map.py and its auxillaries from 18028_Map

JSS95 committed:
Implement classes in trigonometric.py

JSS95 committed:
Introduce process_args

JSS95 committed:
Introduce eval to Sine

JSS95 committed:
Fix typo

JSS95 committed:
Introduce _DeprecatedArgs to map.py

This allows ``args`` of Function and AppliedMap be compatible.
See comment in #18028

JSS95 committed:
Introduce __instancecheck__ to Function

JSS95 committed:
Register trigonometric maps to test_args.py

JSS95 committed:
Update issue number

JSS95 committed:
Update setup.py

mcpl-sympy commented:
@oscarbenjamin @sylee957 
I extracted this part from previous PR. `map/elementary` is yet experimental, so please take a look at other files.  
Especially, I would like to mention that `args` (see sylee's comment in #18028) and `isinstance` compatibility is resolved in this PR.

JSS95 committed:
Introduce eval to Cosine

oscarbenjamin commented:
This is more like what I expected to see. I don't have time to look through it all right now and I'm about to go away for a couple of weeks.

Can you write an explanation in the OP of what the new classes are how, they relate to the old ones, etc.? What new features does the framework potentially add and what issues can be solved?

Why do we need a new package `sympy.map` for this rather than using the existing `sympy.functions`?

One of the most obvious uses for functions as Basic objects is that you can define things like the derivative of a function without reference to any symbol like `f'(0)`. How is differentiation handled in this framework?

As I understand it this doesn't actually change any of the standard sympy functions to the new framework. It would be good (perhaps in a separate experimental PR) to show what changes would be needed to migrate some function class(es) over to this.

There needs to be a way to specify the number of arguments that a function takes. This should be an error:
```
In [7]: Map('f', Reals, Reals)(x, y)                                                                                                           
Out[7]: f(x, y)
```
I think maybe one of parameters to `Map` should be a list like `Map('f', (Reals,), Reals)`. Perhaps we can also allow `Map('f', Reals, Reals)` as a short-hand (make the arg a 1-arg tuple if it isn't already a tuple - `Lambda` does this for its signature).

JSS95 committed:
Change Map to be Basic

Map cannot be evaluated to number

JSS95 committed:
Allow derivative of AppliedMap

mcpl-sympy commented:
@oscarbenjamin 
Oh, the number of arguments can be changed if you pass `ProductSet` to domain. 

```python
>>> Map('f', S.Reals, S.Reals).nargs
1
>>> Map('f', S.Reals**1, S.Reals).nargs
1
>>> Map('f', S.Reals**2, S.Reals).nargs
2
>>> Map('f', S.Reals*S.Complexes, S.Reals).nargs
2
```

I will mention this it in summary.

sylee957 commented:
Although you had devoted a lot of effort of introducing trigonometric functions in the new function, I'd want to hear how much, because there are lots of special function classes in sympy library that can make such work incomplete or exhaustive.

Classifying the properties of functions with respect to the restrictions of underlying sets can give a exhausting amount of work, especially when you see how the sympy functions are defined to behave differently over integers, intervals, reals, complexes, extended reals, ...

mcpl-sympy commented:
@sylee957 
I'm pretty certain that I can migrate every functions in `sympy/functions/elementary` into this framework. Since `args` and `isinstance` are resolved now, hopefully further compatibility issue can be avoided even if some functions are left as `Function`.

JSS95 committed:
Modify docstring of Map to give example for nargs

JSS95 committed:
Introduce derivative.py

Todo: Add docstring and tests

JSS95 committed:
Use abbreviated naming for trig classes

Previous selection was too verbose

JSS95 committed:
Modify test_args.py

JSS95 committed:
Remove _contained logic

Rather, AppliedMap is specially checked by
contains and __contains__ of Set.

JSS95 committed:
Fix test failure

mcpl-sympy commented:
I'm trying to implement function derivative (`f'`), and found out that it cannot be done without introducing `MapAdd`, `MapMul`, and `CompositeMap`. (Without these, I cannot implement `sin' == -cos`)  
Rather than introducing binary operator as I did in my previous commit, which was clearly an overkill, I'm planning to succinctly implement them as classes.

sylee957 commented:
Would this framework be able to replace the usage of `rcall` hacks in the diffgeom module?

mcpl-sympy commented:
@sylee957 Yes - I exactly intend to fix that after this PR.

sylee957 commented:
I’d only note that if things don’t turn out to be realistic scale (Unlike you have confidence that you can replace almost every functions in sympy)

I think that the goal can be changed to replace or providing the parallel implementation for the ‘Function(‘f’)’ that the old sympy was using, and I believe that computing the derivative for specific functions like (sin, cos) can just be tied together via substituting the symbols with the old function framework, which is commonly used lazy evaluation approach that can be used to tie any incompatible computer algebra systems together.

I don’t disagree on your original goal to replace much function symbols, and neither do I agree on the ‘old ways’ that sympy had used to design the functions, but it’s the only remark I can give to make the goals more quick, though I haven’t used to complete long term projects here.

JSS95 committed:
Introduce CompositeMap and IteratedMap

JSS95 committed:
Introduce mapop module

TODO: Add test

JSS95 committed:
Add tests for mapop

JSS95 committed:
Add doctest to derivative.py

_eval_derivative_n_times is added to Map

JSS95 committed:
Make Derivative and DerivativeFunction compatible
with isinstance

JSS95 committed:
Make mapops differentiable

JSS95 committed:
Modify construction signals in mapop.py

JSS95 committed:
Make derivative of unary function printed with prime

JSS95 committed:
Improve derivative logic

Implement fdiff to CompositeMap and IteratedMap
Implement _eval_derivative_n_times to DerivateFunction

JSS95 committed:
Fix Cos.fdiff

JSS95 committed:
Enhance composite function

Composition of multivariate function is now supported

JSS95 committed:
Make compositemap check iterability , not tuple

JSS95 committed:
Make map.rst

JSS95 committed:
Fill map.rst

JSS95 committed:
Un-support partial composition of multivariate function

JSS95 committed:
Fix test failure

eric-wieser commented on sympy/assumptions/handlers/common.py:
```diff
@@ -58,6 +58,21 @@ def Number(expr, assumptions):
+    def Map(expr, assumps):
+        # Unlike other handler, which implies that the object is commutative
+        # with respect to multiplication, this handler implies that this map
+        # commutes the arguments.
```
What does this mean?

mcpl-sympy replied:
With symbol `x` and `y`, `Q.commutative(x)` (or `y`) means that `x*y == y*x`.  
With map `f`, `Q.commutative(f)` means that `f(x, y) == f(y,x)`.

However, I am working to change this because since `MapMul` is introduced, `Q.commutative(f)` must be compatible with `2*f == f*2`. Maybe I'll introduce another handler, `Q.commutative_map`.

JSS95 committed:
Revert _coeff_isneg

JSS95 committed:
Modify assumptions

mcpl-sympy commented:
I am still thinking of implementing this, but it needs some design change. I will close this by now and return with some viable result.
